### PR TITLE
[Packaging] Support Python 3.12

### LINKED ISSUE #27673: Support Python 3.12
## Description
Python 3.12 has been released on 2023-10-02: https://www.python.org/downloads/release/python-3120/

Azure CLI needs to support Python 3.12 like how we supported 3.10 (#19857), 3.11 (#24494).

## TODOs
- [x] [Knack](https://github.com/microsoft/knack) supports Python 3.12
  - https://github.com/microsoft/knack/pull/279
- [x] [azdev](https://github.com/Azure/azure-cli-dev-tools) supports Python 3.12
  - https://github.com/Azure/azure-cli-dev-tools/pull/458
- [x] Dependencies are updated to the Python 3.12 compatible version
  - `pymsalruntime` (dependency of `msal[broker]`) https://github.com/Azure/azure-cli/pull/28556
  - https://github.com/Azure/azure-cli/pull/29466
  - https://github.com/Azure/azure-cli/pull/29486
- [ ] CI passes with Python 3.12
  - Remove `distutils` https://github.com/Azure/azure-cli/pull/28796
  - https://github.com/Azure/azure-cli/pull/27196
  - https://github.com/Azure/azure-cli/pull/29504
  - https://github.com/Azure/azure-cli/pull/29515
  - https://github.com/Azure/azure-cli/pull/29517
  - Remove `assertDictContainsSubset` and `assertNotEquals` #28194
    - https://github.com/Azure/azure-cli/pull/29120
    - https://github.com/Azure/azure-cli/pull/29121
  - https://github.com/Azure/azure-cli/issues/28829
  - https://github.com/Azure/azure-cli/pull/29373
  - https://github.com/Azure/azure-cli/pull/29465
  - https://github.com/Azure/azure-cli/pull/29887
  - https://github.com/Azure/azure-cli-extensions/pull/8010
  - https://github.com/Azure/azure-cli-extensions/pull/8116

### PR DESCRIPTION
**Pull request by bebound**
**Description**

Close #27673

Changes:
- Add 3.12 test
- Add 3.12 support in `setup.py`

### CODE REVIEW

**Comment by yonzhan** [comment]
Packaging

File: src/azure-cli-core/setup.py [hunk 1]
```diff
@@ -39,6 +39,7 @@
     'Programming Language :: Python :: 3.9',
     'Programming Language :: Python :: 3.10',
     'Programming Language :: Python :: 3.11',
+    'Programming Language :: Python :: 3.12',
```
Comment (jiasli) [inline src/azure-cli-core/setup.py:42]: This should also be added to 

https://github.com/Azure/azure-cli/blob/b616adffc346d49fbf851f87dc8497ff28a5ad55/src/azure-cli/setup.py#L48

https://github.com/Azure/azure-cli/blob/a5198b578b17de934e15b1c92e369e45323e9658/src/azure-cli-telemetry/setup.py#L22

https://github.com/Azure/azure-cli/blob/a5198b578b17de934e15b1c92e369e45323e9658/src/azure-cli-testsdk/setup.py#L21

File: .azure-pipelines/breaking-change-tests.yml [hunk 2]
```diff
@@ -22,9 +22,9 @@ jobs:
     fetchTags: true
     persistCredentials: true
   - task: UsePythonVersion@0
-    displayName: 'Use Python 3.10'
+    displayName: 'Use Python 3.11'
```
Comment (bebound) [inline .azure-pipelines/breaking-change-tests.yml:25]: Same as bundled Python version.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_base_decorator.py [hunk 3]
```diff
@@ -24,9 +24,9 @@ def test_validate_decorator_mode(self):
         self.assertEqual(validate_decorator_mode(DecoratorMode.CREATE), True)
         self.assertEqual(validate_decorator_mode(DecoratorMode.UPDATE), True)
         self.assertEqual(validate_decorator_mode(DecoratorMode), False)
-        self.assertEqual(validate_decorator_mode(1), False)
+        # self.assertEqual(validate_decorator_mode(1), False)
```
Comment (bebound) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_base_decorator.py:27]: In 3.12, the emum's value is also `in` Enum.
This is the result in 3.11:
```
>>> from enum import Enum
>>> class DecoratorMode(Enum):
...     """Enumerations used to distinguish whether to handle creation or update.
...     """
...     CREATE = 1
...     UPDATE = 2
...
>>> 1 in DecoratorMode
<stdin>:1: DeprecationWarning: in 3.12 __contains__ will no longer raise TypeError, but will return True or
False depending on whether the value is a member or the value of a member
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\enum.py", line 742, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': 'int' and 'EnumType'

>>> True in DecoratorMode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\enum.py", line 742, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': 'bool' and 'EnumType'

>>> "1" in DecoratorMode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\enum.py", line 742, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': 'str' and 'EnumType'
```
In 3.12
```
>>> 1 in DecoratorMode
True
>>> True in DecoratorMode
True
>>> "1" in  DecoratorMode
False
```

>Changed in version 3.12: Before Python 3.12, a `TypeError` is raised if a non-Enum-member is used in a containment check. --https://docs.python.org/3/library/enum.html#enum.EnumType.__contains__

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_base_decorator.py:27]: Fixing in #29517

**Review by zhoxing-ms — APPROVED** [review]

**Review by jiasli — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Pull request merged**